services/paymaster/src/api.rs

```diff
@@ -51,73 +55,186 @@ pub async fn validate_transaction(
+            for instruction in extra_instructions {
+                if instruction.program_id(transaction.message.static_account_keys())
+                    == &TOLLBOOTH_PROGRAM_ID
+                {
+                    return Err((
+                        StatusCode::BAD_REQUEST,
+                        "Tollbooth instructions must be the first and last instruction only"
+                            .to_string(),
+                    ));
+                }
+            }
+        }
+    } else {
+        // Simulate the transaction to check sponsor SOL spending
+        let simulation_result = rpc
+            .simulate_transaction_with_config(
+                transaction,
+                RpcSimulateTransactionConfig {
+                    sig_verify: false,
+                    replace_recent_blockhash: true,
+                    commitment: Some(CommitmentConfig {
+                        commitment: CommitmentLevel::Processed,
+                    }),
+                    accounts: Some(RpcSimulateTransactionAccountsConfig {
+                        encoding: None,
+                        addresses: vec![sponsor.to_string()],
+                    }),
+                    ..RpcSimulateTransactionConfig::default()
+                },
+            )
+            .map_err(|err| {
                 (
                     StatusCode::INTERNAL_SERVER_ERROR,
-                    format!("Failed to get sponsor balance: {err}"),
+                    format!("Failed to simulate transaction: {err}"),
                 )
             })?;
 
-            let balance_change = pre_balance.saturating_sub(current_balance);
+        if let Some(_err) = simulation_result.value.err {
+            // The paymaster succeeds when the transaction simulation successfully determines that the
+            // transaction returns an error. This is a stopgap measure to unblock 3rd parties while we figure
+            // out the underlying problems with transaction simulation.
+            return Ok(());
+        }
 
-            if balance_change > max_sponsor_spending {
-                return Err((
-                    StatusCode::BAD_REQUEST,
-                    format!(
-                        "Sponsor spending exceeds limit: {balance_change} lamports (max: {max_sponsor_spending})"
-                    ),
-                ));
+        // Check if the sponsor account balance change exceeds the maximum permissible value
+        if let Some(accounts) = simulation_result.value.accounts {
+            if let Some(Some(account)) = accounts.first() {
+                let current_balance = account.lamports;
+                // We need to get the pre-transaction balance to calculate the change
+                let pre_balance = rpc.get_balance(sponsor).map_err(|err| {
+                    (
+                        StatusCode::INTERNAL_SERVER_ERROR,
+                        format!("Failed to get sponsor balance: {err}"),
+                    )
+                })?;
```

Comment:
Shouldn't this be done before simulating in case other txs come through while simulating that change the balance?